### PR TITLE
Fix memory usage of PullParser in D2 builds

### PR DESCRIPTION
**Pull request by mihails-strasuns-sociomantic**
Alternative approach to https://github.com/sociomantic-tsunami/ocean/pull/347

Now `PullParser` operates strictly on slices and does no allocations and
`Document` copies those slices into nodes when necessary.

### CODE REVIEW

File: src/ocean/text/xml/PullParser.d [hunk 1]
```diff
@@ -1,17 +1,13 @@
...
-            Some parts copyright (c) 2009-2016 Sociomantic Labs GmbH.
-            All rights reserved.
-
-        License:
-            Tango Dual License: 3-Clause BSD License / Academic Free License v3.0.
-            See LICENSE_TANGO.txt for details.
-
-        Version: Initial release: February 2008
-
-        Authors: Aaron, Kris
```
Comment (gavin-norman-sociomantic) [inline src/ocean/text/xml/PullParser.d]: I thought we'd deliberately left in the tango authors info?

File: src/ocean/text/xml/PullParser.d [hunk 2]
```diff
@@ -680,10 +670,10 @@ class PullParser(Ch = char)
 
 package struct XmlText(Ch)
 {
-    package Ch*     end;
-    package size_t  len;
-    package Ch[]    text;
-    package Ch*     point;
+    package Ch*      end;
```
Comment (gavin-norman-sociomantic) [inline src/ocean/text/xml/PullParser.d]: Should go in the previous commit.

**Review by gavin-norman-sociomantic — COMMENTED** [review]
I'm not really familiar with the XML parser, so won't comment on the details unless there is a need for an extra opinion, in addition to Ben / Don.

**Comment by don-clugston-sociomantic** [comment]
I think the approach you've taken here is basically right. I don't think it ever makes sense to have a parser that gives mutable slices. The changes to `Document` do mean that it allocates a lot more memory, and does a lot more copying -- every tag in the document will be copied -- but since it allocates one node for every XML element anyway I don't think it's really a problem.
I've never used `Document` but the `PullParser` changes work for me. I will close my PRs on this topic.

I just have one comment, and it's optional.

Since you are cleaning up the formatting of the code, I think you should also delete the untested code in the undocumented `version` statements. There's this horrendous one in Document.d
```d
version(d)
{
                freelists = 0;          // needed to align the codegen!
}
```
I don't know if this is a typo of `version(discrete)` or if they really had `version=d;` (!)

**Review by ben-palmer-sociomantic — APPROVED** [review]
I agree with Dons comments. I also tested with my code and found no problems.

File: src/ocean/text/xml/PullParser.d (hunk 1 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/text/xml/PullParser.d]: I don't remember details but I remember that SVN logs has shown that real authors are quite different from whatever is left in DDOC. And git history is still present anyway. @leandro-lucarella-sociomantic do you know how important is this part legal-wise?

Comment (mihails-strasuns-sociomantic) [reply src/ocean/text/xml/PullParser.d]: (restored just in case)

**Comment by mihails-strasuns-sociomantic** [comment]
Updated. I must say, amount of `version` code that is likely never used in practice worries me a bit.

**Review by don-clugston-sociomantic — APPROVED** [review]
LGTM

**Review by nemanja-boric-sociomantic — APPROVED** [review]
LGTM

**Pull request merged**